### PR TITLE
Feature/png read write

### PR DESCRIPTION
**Pull request by r-englund**
I wrote a reader and writer to load PNG using libpng directly rahter than going through cimg. 
With CImg we only read and write 8 bit even if the pixels are 16. This new module supports 16 bit as well as 8. 

For reading PNG libPNG takes care of converting palettes to RGB/RGBA when needed. Hence, we always get RGB/RGBA Alpha images.
We also support GRAY and GRAY_ALPHA images and outputs them as 1 respectively 2 channel layers with the appropriate swizzle masks.  

When writing float/double images to disk we have to convert them to 16 bit ints (since png only support integers, not floats). This is done right now to using the inviwo function glm_convert_normalized wich gives the same behavior as before. 

**Testing**
For testing a wrote a regression test that uses  images from the [PngSuite](http://www.schaik.com/pngsuite/).  ~160 images of various formats. In our "regular" regression test we have used images generated by inviwo as references, for this regresion test I use the original images as reference. This though does not work well with the imagecompare.py used by the regression test since it requires the image of being of same mode (for example we can't compare a RGB with a Palette)

### CODE REVIEW

File: modules/png/src/pngwriter.cpp [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+
+#include <inviwo/png/pngwriter.h>
+
+#include <inviwo/core/datastructures/image/layer.h>
+#include <inviwo/core/datastructures/image/layerram.h>
+#include <inviwo/core/datastructures/image/layerramprecision.h>
+
+#include <png.h>
+
+#pragma optimize("", off)
```
Comment (petersteneteg) [inline modules/png/src/pngwriter.cpp]: debug?

File: src/core/datastructures/image/layerram.cpp [hunk 2]
```diff
@@ -42,7 +42,7 @@ LayerRAM::LayerRAM(size2_t dimensions, LayerType type, const DataFormatBase* for
 bool LayerRAM::copyRepresentationsTo(LayerRepresentation* targetLayerRam) const {
     static DataWriterType<Layer>* layerWriter_ = InviwoApplication::getPtr()
                                                      ->getDataWriterFactory()
-                                                     ->getWriterForTypeAndExtension<Layer>("png")
+                                                     ->getWriterForTypeAndExtension<Layer>("bmp")
```
Comment (petersteneteg) [inline src/core/datastructures/image/layerram.cpp:48]: This is a horrible hack here... might be nice to add a comment explaining what is going on

File: tools/ivwpy/regression/imagecompare.py [hunk 3]
```diff
@@ -44,11 +44,28 @@ def __init__(self, testImage, refImage, enhance = 10):
         self.numberOfDifferentPixels = None
         self.maxDifference = None
 
+        channels = len(self.testImage.getbands());
+
+
+        # Inviwo don't save PNG using a pallete, if test image is a pallete we
+        # need to convert it to RGB or RGBA
+        if self.refImage.mode == 'P':
```
Comment (petersteneteg) [inline tools/ivwpy/regression/imagecompare.py]: what about 1/2 channels?

File: modules/png/include/inviwo/png/pngreader.h [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+#include <inviwo/core/datastructures/diskrepresentation.h>
+#include <inviwo/core/datastructures/image/layerdisk.h>
+
+namespace inviwo {
+
+    class IVW_MODULE_PNG_API PNGLayerReaderException : public Exception {
+public:
+        PNGLayerReaderException(const std::string& message = "",
+                                ExceptionContext context = ExceptionContext());
+    virtual ~PNGLayerReaderException() throw() {}
```
Comment (petersteneteg) [inline modules/png/include/inviwo/png/pngreader.h]: use noexcept!

File: src/core/datastructures/image/layerram.cpp (hunk 2 above)
Comment (r-englund) [reply src/core/datastructures/image/layerram.cpp:48]: The whole thing with using a DataWriterType<Layer> for resizing/copying is very hacky, but I will add a comment.

File: modules/png/src/pngwriter.cpp (hunk 1 above)
Comment (r-englund) [reply modules/png/src/pngwriter.cpp]: Yes, will remove

File: tools/ivwpy/regression/imagecompare.py (hunk 3 above)
Comment (r-englund) [reply tools/ivwpy/regression/imagecompare.py]: In a PNG image a palette is always representing RGB colors. Alpha can be added as an additional channel. 
Anyway, I changed this a bit, see a9bb185b115f5ba319cf72b1e957ad9dcf88de32

File: tools/ivwpy/regression/app.py [hunk 5]
```diff
@@ -55,14 +55,14 @@ def findModuleTest(path):
...
 	return tests
 
 class App:
-	def __init__(self, 
-		         appPath, 
-				 moduleTestPaths, 
+	def __init__(self,
+		         appPath,
+				 moduleTestPaths,
 				 outputDir,
```
Comment (r-englund) [inline tools/ivwpy/regression/app.py]: Hrm, need to check my space vs tabs settings in my editor

File: src/core/datastructures/image/layerram.cpp (hunk 2 above)
Comment (petersteneteg) [reply src/core/datastructures/image/layerram.cpp:48]: Exactly...

File: modules/png/readme.md [hunk 6]
```diff
@@ -0,0 +1,3 @@
+# png Module
+
+Description of the png module
```
Comment (ResearchDaniel) [inline modules/png/readme.md]: Suggestion: Enables support for reading PNG image files.

File: modules/png/src/pngmodule.cpp [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+#include <inviwo/png/pngmodule.h>
+#include <inviwo/png/pngreader.h>
+#include <inviwo/png/pngwriter.h>
+#include <inviwo/png/pngutils.h>
+
+namespace inviwo {
+
+pngModule::pngModule(InviwoApplication* app) : InviwoModule(app, "png") {
+    
+    LogInfo("Using LibPNG Version " << pngutil::getLibPNGVesrion());
```
Comment (ResearchDaniel) [inline modules/png/src/pngmodule.cpp:39]: Don't think that it is necessary to log this info.

File: modules/png/src/pngreader.cpp [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+    switch (color_type) {
+        case PNG_COLOR_TYPE_GRAY_ALPHA:
+            swizzleMask = swizzlemasks::luminanceAlpha;
+            channels = 2;
+            break;
+        case PNG_COLOR_TYPE_GRAY:
+            swizzleMask = swizzlemasks::luminance;
+            channels = 1;
+            break;
+        case PNG_COLOR_TYPE_PALETTE:
```
Comment (ResearchDaniel) [inline modules/png/src/pngreader.cpp:110]: [[fallthrough]];

File: modules/png/src/pngwriter.cpp [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+        png_set_swap(png_ptr);
+
+        auto pixels = ram->getDataTyped();
+
+        using T2 = typename util::same_extent<T, glm::uint16>::type;
+        std::vector<T2> newData(size.x * size.y);
+        for (int i = 0; i < size.x * size.y; i++) {
+            const static T zero(0);
+            const static T one(1);
+            newData[i] = util::glm_convert_normalized<T2>(glm::clamp(pixels[i], zero, one));
```
Comment (ResearchDaniel) [inline modules/png/src/pngwriter.cpp:116]: Should this not use min/max of the layer data for normalization?

**Review by ResearchDaniel — CHANGES_REQUESTED** [review]

File: modules/png/src/pngwriter.cpp (hunk 9 above)
Comment (r-englund) [reply modules/png/src/pngwriter.cpp:116]: This gives the same behaviour as CImg and will produce the same images as is shown in the canvas. This is used when saving depth textures as pngs, and for the depth textures in thumbnails. 

If we had a datarange/value ranges as we have for volumes we could consider using it here

File: modules/png/src/pngmodule.cpp (hunk 7 above)
Comment (r-englund) [reply modules/png/src/pngmodule.cpp:39]: This log message was in the CImg module before, I just moved it here. The CImg Module has similar log messages for tiff and jpg libs.

**Review by ResearchDaniel — APPROVED** [review]

**Pull request merged**